homeassistant/components/ps4/config_flow.py

```diff
@@ -0,0 +1,124 @@
+"""Config Flow for PlayStation 4."""
+from collections import OrderedDict
+import logging
+
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components.ps4.const import (
+    DEFAULT_NAME, DEFAULT_REGION, DOMAIN, REGIONS)
+from homeassistant.const import (
+    CONF_CODE, CONF_HOST, CONF_IP_ADDRESS, CONF_NAME, CONF_REGION, CONF_TOKEN)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class PlayStation4FlowHandler(config_entries.ConfigFlow):
+    """Handle a PlayStation 4 config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize the config flow."""
+        self.creds = None
+        self.name = None
+        self.host = None
+        self.region = None
+        self.pin = None
+
+    async def async_step_user(self, user_input=None):
+        """Handle a user config flow."""
+        from pyps4_homeassistant import Helper
+        helper = Helper()
+
+        # Abort if device is configured.
+        if self.hass.config_entries.async_entries(DOMAIN):
+            return self.async_abort(reason='devices_configured')
+
+        # Check if able to bind to ports: UDP 987, TCP 997.
+        failed = await self.hass.async_add_executor_job(helper.port_bind)
+        if failed is not None:
+            reason = 'port_{}_bind_error'.format(failed)
```

Comment:
We shouldn't couple the ps4 api and the home assistant api like this. They should be independent.

Please add some logic here that converts the failure returned from the ps4 helper to the reason that we use when aborting.

One way could be a dictionary that maps the failure string to the reason string. If that is constant, put it at the module level.

Reply:
I'm not sure that I follow what you're saying. Can you explain further what you mean by coupling? 

I just call a function in the library. If it fails it returns the port number as an int. This is only so that the port which failed is displayed in the abort message.

Here is the method from the library:
```
    def port_bind(self):
        """Try binding to ports."""
        import socket
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock.settimeout(3.0)
        ports = [987, 997]
        for port in ports:
            try:
                sock.bind(('0.0.0.0', port))
                sock.close()
            except socket.error:
                sock.close()
                return port
            return
```

Reply:
I'm commenting on line 43. That's where the problem is.

We shouldn't use the result from an external api directly in our internal api. Either of those apis might change and we would then break.

We need to define the valid connections between the external api result and the internal api abort reason. Use a dictionary for that mapping.

Reply:
Btw, the external library function returns inconsistently. Always return explicitly a value if any return statement returns a value.

Reply:
So like this:
*UDP_PORT and TCP_PORT are globals
```
    async def async_step_user(self, user_input=None):
        """Handle a user config flow."""
        # Abort if device is configured.
        msg = {
            UDP_PORT: 'port_987_bind_error',
            TCP_PORT: 'port_997_bind_error'}

        ports = msg.keys()

        if self.hass.config_entries.async_entries(DOMAIN):
            return self.async_abort(reason='devices_configured')

        # Check if able to bind to ports: UDP 987, TCP 997.
        failed = await self.hass.async_add_executor_job(
            self.helper.port_bind, ports)
        if failed in ports:
            reason = msg[failed]
            return self.async_abort(reason=reason)
        return await self.async_step_creds(user_input)
```

Reply:
Move msg to a constant at the module level.